cmake: enable lz4 by default

Firefishy opened the pull request:
Enabled building LZ4 compressor support by default.

PR brief discussion with @yuyuyu101 via original LZ4 support https://github.com/ceph/ceph/pull/15434 PR there doesn't seem to be any reason why not to enabled LZ4 by default.

cbodley commented on debian/control:
```diff
@@ -38,6 +38,7 @@ Build-Depends: bc,
                libldap2-dev,
                libleveldb-dev,
                liblttng-ust-dev,
+               liblz4-dev,
```
do we need something similar in ceph.spec.in for rpm builds?

Firefishy replied:
Yup. I'm no RPM expert, I can have a go if desired.

Firefishy replied:
Added lz4-devel RPM build dependency.

tchaikov replied:
could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes, could you prefix it with "debian: " in this case?

theanalyst commented on ceph.spec.in:
```diff
@@ -131,6 +131,7 @@ BuildRequires:	libcurl-devel
 BuildRequires:	libudev-devel
 BuildRequires:	libtool
 BuildRequires:	libxml2-devel
+BuildRequires:	lz4-devel
```
This might be liblz4-devel in rpm distros cc @smithfarm @ktdreyer

smithfarm replied:
In openSUSE/SLE the package really is called `liblz4-devel` as @theanalyst said, and according to http://rpm.pbone.net it's called `lz4-devel` in RHEL/CentOS/Fedora. So this line should be moved down to the "distro-conditional dependencies" section and split into two.

theanalyst commented:
Also can you add a note in PendingReleaseNotes that mentions that LZ4 is also enabled now

Firefishy commented:
Moved the ```liblz4-devel``` / ```lz4-devel``` build dependency to the relevant "distro-conditional dependencies" sections.

Firefishy commented:
@theanalyst Added the PendingReleaseNotes mention, but I'm no wordsmith.

smithfarm commented:
By the way, my assumption here is that this would go into a post-luminous release of Ceph ("M"), not the 12.2.x series. Correct?

theanalyst commented:
jenkins test this please

Firefishy commented:
@smithfarm Personally I'd like to see this in the 12.2.x series too, but I am just a newbie contributor. ;-)

I am unsure to the level of testing lz4 has had.

Firefishy commented:
@theanalyst > jenkins test this please

I am unclear, are you requesting me to add tests for the LZ4 functionality? I suspect that may unfortunately be beyond my ability.

theanalyst commented:
@Firefishy yeah that was for the automated build tests, jenkins failed for an unrelated reason and that invocation triggers the build/tests again

tchaikov commented on PendingReleaseNotes:
```diff
@@ -23,3 +23,6 @@
 
   More details about this port are in: README.FreeBSD
+
+* The inline lz4 compression plugin is now built by default and ships with
```
might want to update `doc/rados/configuration/bluestore-config-ref.rst` accordingly. (search for `lz4` in that file).

tchaikov commented on CMakeLists.txt:
```diff
@@ -272,7 +272,7 @@ endif(WITH_LEVELDB)
 find_package(snappy REQUIRED)
 
-option(WITH_LZ4 "LZ4 compression support" OFF)
+option(WITH_LZ4 "LZ4 compression support" ON)
```
could you prefix the title of your commit message with the subcomponent your are changing ? see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes, in this case, it would be "cmake: ".

tchaikov replied:
```
CMake Error at /usr/share/cmake-3.5/Modules/FindPackageHandleStandardArgs.cmake:148 (message):
  Could NOT find LZ4 (missing: LZ4_LIBRARY LZ4_INCLUDE_DIR) (Required is at
  least version "1.7")
Call Stack (most recent call first):
  /usr/share/cmake-3.5/Modules/FindPackageHandleStandardArgs.cmake:388 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindLZ4.cmake:30 (find_package_handle_standard_args)
  CMakeLists.txt:277 (find_package)
```

again, i don't think we can enable LZ4 by default. for instance, this fails "run-make-check.sh" on trusty.

Firefishy commented:
@tchaikov I've squashed the commits down to a single commit, and removed the bit from the docs about LZ4 not being available in the official releases.

tchaikov commented:
```
In file included from /build/ceph-12.1.2-828-g6dfee5d/src/compressor/lz4/CompressionPluginLZ4.h:21:0,
                 from /build/ceph-12.1.2-828-g6dfee5d/src/compressor/lz4/CompressionPluginLZ4.cc:17:
/build/ceph-12.1.2-828-g6dfee5d/src/compressor/lz4/LZ4Compressor.h: In member function 'virtual int LZ4Compressor::compress(const bufferlist&, ceph::bufferlist&)':
/build/ceph-12.1.2-828-g6dfee5d/src/compressor/lz4/LZ4Compressor.h:33:5: error: 'LZ4_stream_t' was not declared in this scope
     LZ4_stream_t lz4_stream;
```

see https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=trusty,DIST=trusty,MACHINE_SIZE=huge/5617//consoleFull

Firefishy commented:
The build failure is outside of my competency, but I'll try see if I can track it down. I suspect a cmake include issue.

Firefishy commented:
Seem ubuntu 14.04 (trusty) ships with a too old version of liblz4 ( no streaming api).

ktdreyer commented:
Time to drop Trusty?

Firefishy commented:
```cmake``` should likely be taught to disable LZ4 if ```LZ4_VERSION_MAJOR = 1``` and ```LZ4_VERSION_MINOR < 7```

@yaozongyou Any chance you could help with this?

dmick commented:
This also breaks arm64 in its current form; don't merge without addressing that

yuyuyu101 commented:
another option is making lz4 submodule. it will be more ease

tchaikov commented:
@yuyuyu101 it's the last resort. i don't think we should make lz4 yet another submodule just to get it compiled by default. lz4 is stable and is pre-packaged and well-maintained on all well known distros. what is missing is the streaming APIs support on old lz4 packages.

@Firefishy i added the version check to `FindLZ4.cmake` in #17059, but we still cannot enable lz4 on all supported distros releases. unless we can assume that the lz4 with streaming APIs are always available on them.  so i'd suggest close this PR. what do you think?

ktdreyer commented:
Debian bug for updating lz4 https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=860168

tchaikov commented:
@ktdreyer the version packaged by debian is r131, and it has the streaming APIs. it's the r114[1] shipped with ubuntu trusty which does not have the streaming APIs. and xenial packages[2] r131 already.

in short, trusty is too old.

---
[1] https://packages.ubuntu.com/search?keywords=liblz4-dev&searchon=names&suite=trusty&section=all
[2] https://packages.ubuntu.com/search?keywords=liblz4-dev&searchon=names&suite=xenial&section=all

Firefishy commented:
Ideally if #17059 gets merged, I'd then like to amend this PR to have cmake only enable LZ4 if the correct libraries are available. The official rpm / deb builds would not then have LZ4 support enabled, but anyone who installs a recent version of ```liblz4-dev``` / ```lz4-devel``` and builds will get LZ4 support enabled.

ktdreyer commented:
I like that approach @Firefishy

Firefishy commented:
PR amended per my previous comment... Now Ceph LZ4 support is only built if the liblz4 library is manually installed before build. I've removed the LZ4 library from the build requirements for the official RPM / DEB packages.

Firefishy commented:
Failed. Oh well, I tried ;-)

cbodley commented:
> i don't think we can enable LZ4 by default. for instance, this fails "run-make-check.sh" on trusty.

@tchaikov are we still building master on trusty? if not, can we revisit this?

tchaikov commented:
>  are we still building master on trusty? if not, can we revisit this?

no, we are not. sure.

@Firefishy sorry for taking so long, could rebase your change against master and repush?

cbodley commented:
merged in https://github.com/ceph/ceph/pull/21332
